### PR TITLE
[llvm][InstCombine] bitcast bfloat half castpair bug

### PR DESCRIPTION
**Pull request by nasherm**
Miscompilation arises due to instruction combining of cast pairs of the type `bitcast bfloat to half` + `<FPOp> bfloat to half` or `bitcast half to bfloat` + `<FPOp half to bfloat`. For example `bitcast bfloat to half`+`fpext half to double` or `bitcast bfloat to half`+`fpext bfloat to double` respectively reduce to `fpext bfloat to double` and `fpext half to double`. This is an incorrect conversion as it assumes the representation of `bfloat` and `half` are equivalent due to having the same width. As a consequence miscompilation arises.

Fixes #61984

### CODE REVIEW

File: llvm/lib/IR/Instructions.cpp [hunk 1]
```diff
@@ -3214,6 +3214,13 @@ unsigned CastInst::isEliminableCastPair(
         return secondOp;
       return 0;
     case 6:
+      // In cast pairs bfloat and half float shouldn't be treated as equivalent
+      // if the first operation is a bitcast i.e. if we have
+      // bitcast bfloat to half + fpext half to double we shouldn't reduce to
+      // fpext bfloat to double as this isn't equal to fpext half to double.
+      // This has been generalised for all float pairs that have the same width.
+      if (SrcTy->getPrimitiveSizeInBits() == MidTy->getPrimitiveSizeInBits())
```
Comment (ostannard) [inline llvm/lib/IR/Instructions.cpp:3222]: This case is only reached when the first operation is a bitcast, so this condition will always be true. This optimisation should still be valid if SrcTy and MidTy are identical, but I don't know if this code would be reached in that case, or if no-op casts like that are handled elsewhere.

File: llvm/test/Transforms/InstCombine/bitcast-bfloat-half-mixing.ll [hunk 2]
```diff
@@ -0,0 +1,70 @@
+; RUN: opt -S %s | FileCheck %s
+
+define double @F0([2 x bfloat] %P0) {
```
Comment (ostannard) [inline llvm/test/Transforms/InstCombine/bitcast-bfloat-half-mixing.ll:3]: Are the arrays relevant to this test case, or could we use a scalar instead?

File: llvm/test/Transforms/InstCombine/bitcast-bfloat-half-mixing.ll [hunk 3]
```diff
@@ -0,0 +1,70 @@
+; RUN: opt -S %s | FileCheck %s
```
Comment (davemgreen) [inline llvm/test/Transforms/InstCombine/bitcast-bfloat-half-mixing.ll:1]: You probably want to add -passes=instcombine and generating the check lines with the update_test_checks.py script would be good. It keep the tests more standard and maintainable.

File: llvm/test/Transforms/InstCombine/bitcast-bfloat-half-mixing.ll (hunk 2 above)
Comment (nasherm) [reply llvm/test/Transforms/InstCombine/bitcast-bfloat-half-mixing.ll:3]: We can use scalars

File: llvm/test/Transforms/InstCombine/bitcast-bfloat-half-mixing.ll (hunk 3 above)
Comment (nasherm) [reply llvm/test/Transforms/InstCombine/bitcast-bfloat-half-mixing.ll:1]: Cool will do

File: llvm/lib/IR/Instructions.cpp [hunk 4]
```diff
@@ -3214,6 +3214,13 @@ unsigned CastInst::isEliminableCastPair(
         return secondOp;
       return 0;
     case 6:
+      // In cast pairs bfloat and half float shouldn't be treated as equivalent
+      // if the first operation is a bitcast i.e. if we have
+      // bitcast bfloat to half + fpext half to double we shouldn't reduce to
+      // fpext bfloat to double as this isn't equal to fpext half to double.
+      // This has been generalised for all float pairs that have the same width.
+      if(SrcTy->isFloatingPointTy() && MidTy->isFloatingPointTy())
+        return 0;
```
Comment (nikic) [inline llvm/lib/IR/Instructions.cpp:3223]: This implementation doesn't make much sense to me. Note that case 6 is the combination of bitcast + fptoui/fptosi/fpext/fptrunc, so you know that the MidTy is always a float. The below check also ensures that SrcTy is always a float. So this condition in the end is equivalent to "never fold". To implement that, you should replace the entries in the above table with `0` and remove this case entirely.

Comment (nasherm) [reply llvm/lib/IR/Instructions.cpp]: You're right, updated the table in newest patch

File: llvm/lib/IR/Instructions.cpp [hunk 5]
```diff
@@ -3164,7 +3164,7 @@ unsigned CastInst::isEliminableCastPair(
     { 99,99,99, 2, 2,99,99, 8, 2,99,99, 4, 0}, // FPExt          |
     {  1, 0, 0,99,99, 0, 0,99,99,99, 7, 3, 0}, // PtrToInt       |
     { 99,99,99,99,99,99,99,99,99,11,99,15, 0}, // IntToPtr       |
-    {  5, 5, 5, 6, 6, 5, 5, 6, 6,16, 5, 1,14}, // BitCast        |
+    {  5, 5, 5, 0, 0, 5, 5, 0, 0,16, 5, 1,14}, // BitCast        |
```
Comment (nikic) [inline llvm/lib/IR/Instructions.cpp:3167]: Can you please also remove case 6 from the below switch?

**Comment by nasherm** [comment]
clang-format expects entire `CastResults` table to be formatted in a certain way, but this formatting affects readability

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**